### PR TITLE
#842 Forbid too long call chains

### PR DESCRIPTION
**Pull request by lucaionescu**
# I have made things!

Hi, the implementation of the `CallChainsVisitor` is not correct yet, I just wanted to open the PR as work in progress for me to keep track of it and for you to be able add comments to it.

Thanks!

## Checklist

- [x] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [x] I have created at least one test case for the changes I have made
- [x] I have updated the documentation for the changes I have made
- [x] I have added my changes to the `CHANGELOG.md`

### CODE REVIEW

File: wemake_python_styleguide/visitors/ast/complexity/calls.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            chained_item = self._chained_item(iterator)
+            if chained_item is None:
+                return
+            iterator = chained_item
+
+    def _check_consecutive_call_number(self, node: ast.Call) -> None:
+        if node in self._visited_calls:
+            return
+
+        consecutive_calls = cast(Set[ast.Call], set(takewhile(
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/complexity/calls.py]: Can you please remove `cast`? We need to refactor this part.

**Comment by sobolevn** [comment]
@lucaionescu can you please rebase your changes?

**Comment by sobolevn** [comment]
Oops, something went wrong! There are now unrelated changes in the PR: https://github.com/wemake-services/wemake-python-styleguide/pull/945/files#diff-6a3371457528722a734f3c51d9238c13R106

**Comment by lucaionescu** [comment]
Sorry for that. What would be the best way to proceed now? Should I remove these extra files from the PR, revert the changes and do a clean rebase?

**Comment by sobolevn** [comment]
@lucaionescu no problem 👍 

Yes, I guess that is the best solution. I sometimes use `rm -rf` and copy paste stuff 😄

**Comment by lucaionescu** [comment]
Okay, that was a wild ride, but I feel like I learned a ton of new Git stuff. 

Also, when I run `make lint`, I get the following error:
```
flake8 .
Traceback (most recent call last):
  File "/Users/ioan/Library/Caches/pypoetry/virtualenvs/wemake-python-styleguide-py3.7/lib/python3.7/site-packages/flake8/plugins/manager.py", line 158, in load_plugin
    self._load()
  File "/Users/ioan/Library/Caches/pypoetry/virtualenvs/wemake-python-styleguide-py3.7/lib/python3.7/site-packages/flake8/plugins/manager.py", line 135, in _load
    self._plugin = self.entry_point.load()
  File "/Users/ioan/Library/Caches/pypoetry/virtualenvs/wemake-python-styleguide-py3.7/lib/python3.7/site-packages/entrypoints.py", line 82, in load
    mod = import_module(self.module_name)
  File "/Users/ioan/miniconda3/envs/wemake-dev/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/Users/ioan/Documents/wemake-python-styleguide/wemake_python_styleguide/checker.py", line 55, in <module>
    from wemake_python_styleguide.presets.types import tree as tree_preset
  File "/Users/ioan/Documents/wemake-python-styleguide/wemake_python_styleguide/presets/types/tree.py", line 5, in <module>
    from wemake_python_styleguide.presets.topics import complexity
  File "/Users/ioan/Documents/wemake-python-styleguide/wemake_python_styleguide/presets/topics/complexity.py", line 5, in <module>
    from wemake_python_styleguide.visitors.ast.complexity import (
  File "/Users/ioan/Documents/wemake-python-styleguide/wemake_python_styleguide/visitors/ast/complexity/function.py", line 9, in <module>
    from wemake_python_styleguide.logic import complexity, functions
  File "/Users/ioan/Documents/wemake-python-styleguide/wemake_python_styleguide/logic/complexity.py", line 3, in <module>
    from cognitive_complexity.api import get_cognitive_complexity
ModuleNotFoundError: No module named 'cognitive_complexity'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/ioan/Library/Caches/pypoetry/virtualenvs/wemake-python-styleguide-py3.7/bin/flake8", line 10, in <module>
    sys.exit(main())
  File "/Users/ioan/Library/Caches/pypoetry/virtualenvs/wemake-python-styleguide-py3.7/lib/python3.7/site-packages/flake8/main/cli.py", line 18, in main
    app.run(argv)
  File "/Users/ioan/Library/Caches/pypoetry/virtualenvs/wemake-python-styleguide-py3.7/lib/python3.7/site-packages/flake8/main/application.py", line 393, in run
    self._run(argv)
  File "/Users/ioan/Library/Caches/pypoetry/virtualenvs/wemake-python-styleguide-py3.7/lib/python3.7/site-packages/flake8/main/application.py", line 380, in _run
    self.initialize(argv)
  File "/Users/ioan/Library/Caches/pypoetry/virtualenvs/wemake-python-styleguide-py3.7/lib/python3.7/site-packages/flake8/main/application.py", line 363, in initialize
    self.find_plugins()
  File "/Users/ioan/Library/Caches/pypoetry/virtualenvs/wemake-python-styleguide-py3.7/lib/python3.7/site-packages/flake8/main/application.py", line 199, in find_plugins
    self.check_plugins.load_plugins()
  File "/Users/ioan/Library/Caches/pypoetry/virtualenvs/wemake-python-styleguide-py3.7/lib/python3.7/site-packages/flake8/plugins/manager.py", line 410, in load_plugins
    plugins = list(self.manager.map(load_plugin))
  File "/Users/ioan/Library/Caches/pypoetry/virtualenvs/wemake-python-styleguide-py3.7/lib/python3.7/site-packages/flake8/plugins/manager.py", line 297, in map
    yield func(self.plugins[name], *args, **kwargs)
  File "/Users/ioan/Library/Caches/pypoetry/virtualenvs/wemake-python-styleguide-py3.7/lib/python3.7/site-packages/flake8/plugins/manager.py", line 408, in load_plugin
    return plugin.load_plugin()
  File "/Users/ioan/Library/Caches/pypoetry/virtualenvs/wemake-python-styleguide-py3.7/lib/python3.7/site-packages/flake8/plugins/manager.py", line 165, in load_plugin
    raise failed_to_load
flake8.exceptions.FailedToLoadPlugin: Flake8 failed to load plugin "WPS" due to No module named 'cognitive_complexity'.
make: *** [lint] Error 1
```

**Comment by sobolevn** [comment]
You should run `poetry install`, we have some new dependencies since your last installation.

Thanks a lot for your time and effort! I genuinely appreciate it.

**Comment by lucaionescu** [comment]
`poetry install`resolved those issues, thanks!
Sorry for working on it so slowly, my semester recently started and I am involved in many things, but I really enjoy working on this. 

Can you please have a look at `wemake_python_styleguide/logic/calls.py`? There is a logical error in there, but I can't figure it out. When I run the tests they remain "stuck" when coming across testing this violation.

**Comment by sobolevn** [comment]
@lucaionescu no worries, please work as much as it comfortable for you. 

Thanks a lot for your time and effort. I really appreciate each community contribution.

**Comment by sobolevn** [comment]
Thanks a lot! I have adapted several things for the current code base and merged your changes. 👍

**Comment by lucaionescu** [comment]
Of course, thank you for allowing me to contribute!

**Pull request merged**